icefoganalytics/internal-data-portal
# System Admin: Can Add/Edit/Delete Users

Fixes https://github.com/icefoganalytics/internal-data-portal/issues/66

Relates to:
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)

# Context

Give admins the ability to create/update/delete users.

# Implementation

Add an "all users" page with a paginated table.
Make edit button redirect to edit page for user.
Make create button redirect to users/new page.
Make delete icon open a confirmation modal.
Implement user creation as a process that sets the `auth0_subject` to null and sets the new `setup_from_email_first_login` flag on the user. If a user logs in with an email matching the newly created user, it will updated the record, using the `auth0_subject` from the user, and unsetting the `setup_from_email_first_login` flag.

NOTE: did not create a "user" read page, though that will likely be useful in the future.

# Screenshots

All users page
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/aabbd6b0-8df3-4946-bae3-f96dfde1d40f)

Edit user page
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/dc94e175-37e2-49fc-9ea0-ca713e615e24)

Delete user modal
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/4a053b1e-58a4-4b1f-85d4-52cf0048fb93)

Create user page
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/78cc014c-ef0f-490f-9cdb-0a3b2f8f5808)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Using the QA helpers, make yourself a system admin user.
5. Check that you can now see an "all users" entry in the top nav kebab menu.
6. Click on the link and check that you can now view info about all users in a paginated table.
7. Click the edit button to edit a user.
8. Check that the breadcrumbs allow navigation back to the "all users" page.
9. Check that the back and cancel buttons both redirect back to the "all users" page.
10. Check that you can edit the user. Note that save is so fast, that it might not even be detectable. I'm might be worth setting up auto-save here and on the user profile page?
11. Check that you can trigger a "sync profile with external directory" for a given user.
12. Check that you can delete a user.
13. Check that you can create a new user. Give the user an email that you control and test signup to that email.
14. At the very end, go back to the dashboard and become a non-system admin user.
15. Check that the "all users" top nav kebab menu is now gone.
